BlackCubes/react-challenge-project-sep-trial
# API Call Refactor

## Changes
1. Separated and refactored any APIs in `orderActions.js` and `authActions.js` into separate files of `orderAPI.js` and `authAPI.js`.
2. Refactored `actions/types.js` into different files of `constants/orderActions.js` and `constants/authActions.js` and some of their names.
3. Added a loading feature to redux for "orders" and "auth".
4. Disabled certain form implementations when the loading state is active.
5. Created a persistor with `redux-persist`.

## Purpose
There should be refactors on the code to leverage Redux to handle API calls, and store the results inside Redux so any component can have access to them, and adjust the components accordingly.

## Approach
The first approach was to separate any API calls in `orderActions.js` into its own file called `orderAPI.js` which would be stored in a folder called `api`. This was also done with any API calls in `authActions.js`, but to its own file called `authAPI.js`. Both files utilizes the `axios` package along with a custom `headers` utility function to store headers and in the future, any JWTs that needs to be sent. Lastly, `orderActions.js` and `authActions.js` were refactored with these new API files. Lastly, the reason why the `api` folder is stored outside of the `redux` folder is separation of concerns: Anything related to Redux should be in the `redux` folder and API calls does not have any Redux. It just calls the APIs.

Next, the constant action type variables in `types.js` were also put into a new folder called `constants` in the `redux` folder, and were separated into two files: `authTypes.js` for any auth, and `orderTypes.js` for any orders. This helps to better stay organize between different reducers.

Next, a new feature in both the auth and order actions was to include "loading". New actions, action types, and reducers were created for this "loading" feature. Lastly, the form functionalities were implemented using "loading" for auth and orders.

A tiny bug fix was done for the `ProtectedRoute` component that changes the path for any authenticated users to be redirected to the `/view-orders` path and not to the homepage. This was done because when the user logs in, the `/login` path is for any non-authenticated users which would redirect the user to the homepage once the user logs in successfully based on the `ProtectedRoute` component.

The final feature was implementing `redux-persist` which persists and rehydrates the redux store. What this means is that whenever the user refreshes the page or uses the forward/back buttons, the redux store would not get erased back to its initial state since React is a Single-Page Application. The way `redux-persist` does this is by storing the redux store into `localStorage`. Now, whenever the user refreshes, the redux store does not reset. Because of this new implementation, any custom `localStorage` to store the user's email and token was removed in the redux actions, and any usage of it in the components are also removed.

## Testing
1. Pull changes.
2. On the parent directory, run `docker compose down && docker compose up --build` on your terminal.
3. On the `/login` path, login and make sure you are redirected to the `/view-orders` path.
4. Create an order in the `/order` path and go back to the `/view-orders` path.
5. Refresh the page and make sure your email in the navbar nor the order(s) you created are erased.

## Learning
[Redux Persist](https://www.npmjs.com/package/redux-persist)

Closes #9
